### PR TITLE
Implement user actions for font resize

### PR DESCRIPTION
**Pull request by da-x**

### CODE REVIEW

File: alacritty.yml [hunk 1]
```diff
@@ -185,6 +185,8 @@ key_bindings:
   - { key: Q,        mods: Command, action: Quit                         }
   - { key: W,        mods: Command, action: Quit                         }
   - { key: Insert,   mods: Shift,   action: PasteSelection               }
+  - { key: Equals,   mods: Control, action: IncreaseFontSize             }
+  - { key: Subtract, mods: Control, action: DecreaseFontSize             }
```
Comment (maximbaz) [inline alacritty.yml]: make sure to replicate these changes in `alacritty_macos.yml` as well :+1:

**Comment by siiptuo** [comment]
I think that `key: Add` should be added as another default for `IncreaseFontSize`. For reference, GNOME Terminal seems to only support `Ctrl`+`+` but Konsole and many GUI applications also support `Ctrl`+`=`. Different keyboard layouts may require multiple modifier keys that are interpreted incorrectly in Alacritty. For instance in US layout it's not possible to enter `Ctrl`+`+` by pressing keys `Ctrl`, `Shift` and `=`.

**Comment by inducer** [comment]
Fantastic, works like a charm with current master! Thank you so much for your work on this. FWIW, the merge conflicts are easily resolved.

**Comment by inducer** [comment]
One remaining nicety might be a `ResetFontSize` action (Ctrl-0 in gnome-terminal).

**Comment by da-x** [comment]
Rebased, fixed, and extended for a Reset action.

**Comment by mhartington** [comment]
Applied this patch, but having an issue with trying to bind to `Subtract`

```
{ key: Subtract, mods: Control, action: DecreaseFontSize             }
```

It instead will just print out the character `-` instead of performing the action.

**Comment by athaeryn** [comment]
@mhartington Changing `Subtract` to `Minus` fixed it for me. :)

**Comment by mhartington** [comment]
Perfect! Thanks @athaeryn

**Comment by da-x** [comment]
Perhaps we should have both Subtract and Minus mapped for the decrease action
@jwilm

**Comment by lstyls** [comment]
@jwilm @da-x: Friendly ping. I'm happy to patch this, but I think a lot of new users would be delighted if this were landed.

**Comment by da-x** [comment]
I've added the ability to switch between fonts automatically based on the current font size, picking a bitmap font for smaller sizes and anti-aliased TTF for the larger sizes. Available in my fork: https://github.com/da-x/alacritty/tree/auto-small-font

**Comment by uri** [comment]
Is this PR waiting on something? I've been using `da-x:font-resize` a while now with no problems.

**Comment by mishchea** [comment]
@jwilm any update on the status of this PR? This is the single feature that is keeping me from using alacritty...

**Comment by norcalli** [comment]
I have a small suggestion, which is to allow specifying the font delta step size in the config, or perhaps making it `0.5` in case that's too much work, since I often work in half step differences because my monitor is HiDPI.

Also @jwilm either way I think this PR has been baked in for a while and it would be a handy change. It doesn't seem like it's been given any attention yet, so perhaps even an explanation of why it's not been merged would be sufficient.

**Comment by mathstuf** [comment]
> It doesn't seem like it's been given any attention yet, so perhaps even an explanation of why it's not been merged would be sufficient.

On another PR or issue, @jwilm mentioned that he's traveling and may have time to work on it after he is done with it. He also mentioned an interest in getting the project down to 0 open PRs, so this should get attention eventually.

**Comment by jwilm** [comment]
Thanks @da-x for building this, everyone who has tested the feature, and thanks to everybody for your patience on landing it.

I just realized that this has a bit of a memory leak that I'm going to work on patching right now.

**Comment by jwilm** [comment]
The leak should be fixed on master.

**Comment by da-x** [comment]
BTW, why the `#![feature(alloc_system)] extern crate alloc_system` change? It needs non-stable Rust for building, unfortunately. If I remove that line it builds on stable without a hitch.

```
error[E0554]: #![feature] may not be used on the stable release channel
  --> src/main.rs:18:1
   |
18 | #![feature(alloc_system)] extern crate alloc_system;
```

**Comment by jwilm** [comment]
My bad, I accidentally left that in after tracking down the leak.

**Comment by da-x** [comment]
Thanks.

BTW what do you think about the alternative small font feature ( https://github.com/da-x/alacritty/commit/3d6b874907491d301ab86a7a264957bf8e7caed7 ). Should open a PR?

**Comment by NickeZ** [comment]
This doesn't work for me.

I updated the config with:
```
  - { key: Key0,     mods: Control, action: ResetFontSize                }
  - { key: Equals,   mods: Control, action: IncreaseFontSize             }
  - { key: Subtract, mods: Control, action: DecreaseFontSize             }
```
But it prints an equals sign `=` instead of changing the size of the font..

I attach a short log when I press Ctrl+`=`.
[log.txt](https://github.com/jwilm/alacritty/files/1391007/log.txt)

**Comment by mathstuf** [comment]
FWIW, it works for me, but I use `Shift+{Left,Up,Down}` instead.

**Comment by unphased** [comment]
What's the deal with `Subtract` vs `Minus`? Why does it not follow http://tomaka.github.io/glium_text/glium/glutin/enum.VirtualKeyCode.html ?

**Comment by mathstuf** [comment]
`Subtract` is also there. One might be for the numpad? I'm not seeing docs distinguishing the two.

**Comment by unphased** [comment]
binding to `Subtract` flat-out doesn't work on OS X here (10.11.6). Seems like you have to use Minus. Which is interesting because here's an error message which I saw when I erroneously set the config with `x` (lowercase) once, and it spit out: 

```Ignoring invalid config: problem with config: key_bindings[2].key: unknown variant `x`, expected one of `Key1`, `Key2`, `Key3`, `Key4`, `Key5`, `Key6`, `Key7`, `Key8`, `Key9`, `Key0`, `A`, `B`, `C`, `D`, `E`, `F`, `G`, `H`, `I`, `J`, `K`, `L`, `M`, `N`, `O`, `P`, `Q`, `R`, `S`, `T`, `U`, `V`, `W`, `X`, `Y`, `Z`, `Escape`, `F1`, `F2`, `F3`, `F4`, `F5`, `F6`, `F7`, `F8`, `F9`, `F10`, `F11`, `F12`, `F13`, `F14`, `F15`, `Snapshot`, `Scroll`, `Pause`, `Insert`, `Home`, `Delete`, `End`, `PageDown`, `PageUp`, `Left`, `Up`, `Right`, `Down`, `Back`, `Return`, `Space`, `Compose`, `Numlock`, `Numpad0`, `Numpad1`, `Numpad2`, `Numpad3`, `Numpad4`, `Numpad5`, `Numpad6`, `Numpad7`, `Numpad8`, `Numpad9`, `AbntC1`, `AbntC2`, `Add`, `Apostrophe`, `Apps`, `At`, `Ax`, `Backslash`, `Calculator`, `Capital`, `Colon`, `Comma`, `Convert`, `Decimal`, `Divide`, `Equals`, `Grave`, `Kana`, `Kanji`, `LAlt`, `LBracket`, `LControl`, `LMenu`, `LShift`, `LWin`, `Mail`, `MediaSelect`, `MediaStop`, `Minus`, `Multiply`, `Mute`, `MyComputer`, `NavigateForward`, `NavigateBackward`, `NextTrack`, `NoConvert`, `NumpadComma`, `NumpadEnter`, `NumpadEquals`, `OEM102`, `Period`, `PlayPause`, `Power`, `PrevTrack`, `RAlt`, `RBracket`, `RControl`, `RMenu`, `RShift`, `RWin`, `Semicolon`, `Slash`, `Sleep`, `Stop`, `Subtract`, `Sysrq`, `Tab`, `Underline`, `Unlabeled`, `VolumeDown`, `VolumeUp`, `Wake`, `WebBack`, `WebFavorites`, `WebForward`, `WebHome`, `WebRefresh`, `WebSearch`, `WebStop`, `Yen` at line 245 column 12```

This set includes both `Minus` and `Subtract` so yes maybe numlock has something to do with it. With that being said, is NumLock even a concept at all on Macs??

**Comment by mathstuf** [comment]
> is NumLock even a concept at all on Macs?

Yes, but only if you don't use Apple keyboards.

**Comment by 22459** [comment]
I'd like to bind my font resize actions to control-scrollup/down, could something be implemented for this?

**Comment by mathstuf** [comment]
@22459 Yes, I have this in my config:

```yaml
  - { key: Left,     mods: Shift,   action: ResetFontSize                }
  - { key: Up,       mods: Shift,   action: IncreaseFontSize             }
  - { key: Down,     mods: Shift,   action: DecreaseFontSize             }
```

**Comment by mathstuf** [comment]
Oh, I read *pageup* and *pagedown*. I think the mouse buttons are 4 and 5 for the scrollwheel. See [here](https://github.com/jwilm/alacritty/blob/master/alacritty.yml#L191).

**Comment by 22459** [comment]
The problem here is also that it's a mouse/keyboard combination?

**Comment by 22459** [comment]
Also I don't think that has access to the keyboard actions?

**Comment by mathstuf** [comment]
There's `mouse_bindings` which takes an `action` value as well. It probably does just as well with the `FontSize` actions.

**Comment by mabushey** [comment]
Is this ever going to be implemented?

**Comment by blakehawkins** [comment]
Since this PR was merged, the config file is now TOML.

The key names are all the same. I've used Minus instead of Subtract because Subtract is not recognised on my platform/keyboard/?

```
[keyboard]
bindings = [
  { key = "Equals", mods = "Control", action = "IncreaseFontSize" },
  { key = "Minus", mods = "Control", action = "DecreaseFontSize" },
]
```

**Comment by kchibisov** [comment]
All these bindings are set by default, so no need to set any of them yourself.

**Comment by blakehawkins** [comment]
These did not work for me by default

**Comment by kchibisov** [comment]
with empty config they should, on macOS though you should use `command`, but please open a proper issue and not necrobump PRs.

**Pull request merged**